File: automation-test-playwright/pages/customer/CustomerContractListPage.ts
import { type Locator, type Page } from "@playwright/test";
import { CrudListPage } from "../core/CrudListPage";

export class CustomerContractListPage extends CrudListPage {
  protected readonly path = "/customer/contract/list";
  readonly list: Locator;

  /**
   * Initializes this page object.
   */
  constructor(page: Page) {
    super(page);
    this.list = this.page.locator("#contractList");
  }

  async waitForLoaded(): Promise<void> {
    await this.page.waitForURL(/\/customer\/contract\/list|\/customer\/contracts/);
    await this.waitForTitleContainsLoose("hop dong", "contract");
    await this.waitForVisible(this.list);
  }

  async waitForContractData(): Promise<void> {
    await this.waitForCondition(async () => {
      const hasCards = (await this.page.locator("#contractList .contract-container").count()) > 0;
      const hasEmpty = await this.page
        .locator("#contractList .empty-state")
        .isVisible()
        .catch(() => false);
      return hasCards || hasEmpty;
    }, "Contract list did not load cards or empty state.");
  }

  /**
   * Applies the building filter.
   */
  async filterByBuilding(buildingId: number | string): Promise<void> {
    await this.selectFilter("buildingId", String(buildingId));
  }

  /**
   * Applies the status filter.
   */
  async filterByStatus(status: "ACTIVE" | "EXPIRED"): Promise<void> {
    await this.selectFilter("status", status);
  }

  /**
   * Submits the active filters.
   */
  async submitFilters(): Promise<void> {
    await this.search();
  }

  cardByBuildingName(name: string): Locator {
    return this.firstVisible(this.page.locator("#contractList .contract-container").filter({ hasText: name }));
  }

  async waitForEmptyState(): Promise<void> {
    await this.waitForVisible(this.page.locator("#contractList .empty-state"));
  }
}

File: automation-test-selenium/src/main/java/org/suntower/pages/customer/CustomerContractListPage.java
package org.suntower.pages.customer;

import org.openqa.selenium.JavascriptExecutor;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.WebElement;
import org.openqa.selenium.support.ui.ExpectedConditions;
import org.suntower.pages.core.CrudListPage;

public class CustomerContractListPage extends CrudListPage {
  private final org.openqa.selenium.By list = css("#contractList");

  public CustomerContractListPage(WebDriver driver) {
    super(driver);
    this.path = "/customer/contract/list";
  }

  public void waitForLoaded() {
    waitForUrlMatches(".*/customer/(contract/list|contracts).*");
    waitForVisible(list);
  }

  public void waitForContractData() {
    waitForCondition(
        () -> !driver.findElements(css("#contractList .contract-container")).isEmpty() || isVisible(css("#contractList .empty-state")),
        "Customer contract list did not load cards or empty state.");
  }

  public void filterByBuilding(long buildingId) {
    setFormControlValue("buildingId", String.valueOf(buildingId));
  }

  public void filterByStatus(String status) {
    setFormControlValue("status", status);
  }

  public void submitFilters() {
    wait.until(ExpectedConditions.presenceOfElementLocated(css("#filterForm")));
    ((JavascriptExecutor) driver).executeScript("document.querySelector('#filterForm').requestSubmit();");
    waitForContractData();
  }

  public WebElement cardByBuildingName(String name) {
    waitForCondition(
        () ->
            driver.findElements(css("#contractList .contract-container")).stream()
                .anyMatch(
                    card -> {
                      try {
                        return card.isDisplayed() && card.getText().contains(name);
                      } catch (RuntimeException ignored) {
                        return false;
                      }
                    }),
        "No visible customer contract card contained text: " + name);
    return driver.findElements(css("#contractList .contract-container")).stream()
        .filter(
            card -> {
              try {
                return card.isDisplayed() && card.getText().contains(name);
              } catch (RuntimeException ignored) {
                return false;
              }
            })
        .findFirst()
        .orElseThrow();
  }

  public void waitForEmptyState() {
    waitForVisible(css("#contractList .empty-state"));
  }

  private void setFormControlValue(String fieldName, String value) {
    wait.until(ExpectedConditions.presenceOfElementLocated(inputByName(fieldName)));
    ((JavascriptExecutor) driver)
        .executeScript(
            "const el = document.querySelector('[name=\"' + arguments[0] + '\"]');"
                + "if (el.tagName === 'SELECT' && !Array.from(el.options).some(option => option.value === arguments[1])) { el.add(new Option(arguments[1], arguments[1])); }"
                + "el.value = arguments[1];"
                + "el.dispatchEvent(new Event('input', { bubbles: true }));"
                + "el.dispatchEvent(new Event('change', { bubbles: true }));",
            fieldName,
            value);
  }
}
